# Read the raw data.

In [1]:
import pandas as pd

# Define the folder path
data_path = "/home/er_bim/Indonesia_COE/data/raw/"

In [2]:
# Read the data from local and define them as variables

# Year 2021
coffee21 = pd.read_csv(data_path + 'indonesia_2021_coffee.csv', index_col=0)
auction21 = pd.read_csv(data_path + 'indonesia_2021_auction.csv', index_col=0)

# Year 2022
coffee22 = pd.read_csv(data_path + 'indonesia_2022_coffee.csv', index_col=0)
auction22 = pd.read_csv(data_path + 'indonesia_2022_auction.csv', index_col=0)

# Year 2023
coffee23 = pd.read_csv(data_path + 'indonesia_2023_coffee.csv', index_col=0)
auction23 = pd.read_csv(data_path + 'indonesia_2023_auction.csv', index_col=0)

# Data Wrangling.

In this section, we will do the data cleaning and manipulation that is required, based on the basic data structure.

## Column names consistency handling.

The tables are grouped as the 'coffees' and 'auctions'.

In [3]:
# Create a list of DataFrames
coffees = [coffee21, coffee22, coffee23]
strings = ['coffee21', 'coffee22', 'coffee23']

# Checking columns names in coffees group
for c, s in zip (coffees, strings):
    cols = c.columns
    print(f"The columns in {s} are: \n{cols}\n\n")

The columns in coffee21 are: 
Index(['RANK', 'SCORE', 'FARM', 'FARMER', 'REGION', 'WEIGHT (kg)', 'VARIETY',
       'PROCESS'],
      dtype='object')


The columns in coffee22 are: 
Index(['Rank', 'Score', 'Farm', 'Farmer', 'Region', 'Weight(Lbs)', 'Variety',
       'Process'],
      dtype='object')


The columns in coffee23 are: 
Index(['Rank', 'Score', 'Farmer Name', 'Farm Name', 'Region',
       'Estimated Weight (kg)', 'Process', 'Variety'],
      dtype='object')




In [4]:
# Create a list of DataFrames
auctions = [auction21, auction22, auction23]
texts = ['auction21', 'auction22', 'auction23']

# Checking columns names in auctions group
for a, t in zip (auctions, texts):
    cols = a.columns
    print(f"The columns in {t} are: \n{cols}\n\n")

The columns in auction21 are: 
Index(['Rank', 'Farm', 'Score', 'Weight (lbs)', 'High Bid', 'Total Value',
       'Company Name'],
      dtype='object')


The columns in auction22 are: 
Index(['Rank', 'Farm', 'Score', 'Weight (Lbs)', 'High Bid', 'Total Value',
       'Company Name'],
      dtype='object')


The columns in auction23 are: 
Index(['Ranking', 'Score', 'Farm', 'Weight (lb)', 'Price per lb',
       'Total Price', 'Buyer'],
      dtype='object')




We find some inconsistencies in the columns nomenclature, and the uppercase characters in the column names of "coffee21" data.

In general the year 2023 tables have different columns names compare to the previous 2 years, although they have the same definition.

We will handle them later.

### Convert the uppercase column names.

There is one data which contains the column names in uppercase format, we will convert it to the title case to make it consistent.

In [5]:
# Check the column names in coffee21
coffee21.columns

Index(['RANK', 'SCORE', 'FARM', 'FARMER', 'REGION', 'WEIGHT (kg)', 'VARIETY',
       'PROCESS'],
      dtype='object')

In [6]:
# Convert the column names in coffee21 into title case
coffee21.columns = coffee21.columns.str.title()

In [7]:
# Recheck the column names in coffee21
coffee21.columns

Index(['Rank', 'Score', 'Farm', 'Farmer', 'Region', 'Weight (Kg)', 'Variety',
       'Process'],
      dtype='object')

### Inconsistent column names terming correction.

Let's recheck the column names for all data.

In [8]:
# Rechecking columns names in coffees group
for c, s in zip (coffees, strings):
    cols = c.columns
    print(f"The columns in {s} are: \n{cols}\n\n")

The columns in coffee21 are: 
Index(['Rank', 'Score', 'Farm', 'Farmer', 'Region', 'Weight (Kg)', 'Variety',
       'Process'],
      dtype='object')


The columns in coffee22 are: 
Index(['Rank', 'Score', 'Farm', 'Farmer', 'Region', 'Weight(Lbs)', 'Variety',
       'Process'],
      dtype='object')


The columns in coffee23 are: 
Index(['Rank', 'Score', 'Farmer Name', 'Farm Name', 'Region',
       'Estimated Weight (kg)', 'Process', 'Variety'],
      dtype='object')




In [9]:
# Checking columns names in auctions group
for a, t in zip (auctions, texts):
    cols = a.columns
    print(f"The columns in {t} are: \n{cols}\n\n")

The columns in auction21 are: 
Index(['Rank', 'Farm', 'Score', 'Weight (lbs)', 'High Bid', 'Total Value',
       'Company Name'],
      dtype='object')


The columns in auction22 are: 
Index(['Rank', 'Farm', 'Score', 'Weight (Lbs)', 'High Bid', 'Total Value',
       'Company Name'],
      dtype='object')


The columns in auction23 are: 
Index(['Ranking', 'Score', 'Farm', 'Weight (lb)', 'Price per lb',
       'Total Price', 'Buyer'],
      dtype='object')




We can see that some column name terms are different, even though they have the same definition.

Now we alter those different column names to match them with the previous years'.

In [10]:
# Define the dictionary for column names replacements
cof = {'Farm Name': 'Farm', 'Farmer Name': 'Farmer'}
auc = {'Ranking': 'Rank', 'Weight (lb)': 'Weight (Lbs)', 'Weight (lbs)': 'Weight (Lbs)',\
    'Price per lb': 'High Bid', 'Total Price': 'Total Value', 'Company Name': 'Buyer'}

In [11]:
# Convert the column names in auction2021
auction21.rename(columns = auc, inplace = True)

# Convert the column names in auction2022
auction22.rename(columns = auc, inplace = True)

# Convert the column names in auction2023
auction23.rename(columns = auc, inplace = True)

# Convert the column names in coffee2023
coffee23.rename(columns = cof, inplace = True)

In [12]:
# Updating the coffees group data
coffees = coffee21, coffee22, coffee23

# Rechecking columns names in coffees group
for c, s in zip (coffees, strings):
    cols = c.columns
    print(f"The columns in {s} are: \n{cols}\n\n")

The columns in coffee21 are: 
Index(['Rank', 'Score', 'Farm', 'Farmer', 'Region', 'Weight (Kg)', 'Variety',
       'Process'],
      dtype='object')


The columns in coffee22 are: 
Index(['Rank', 'Score', 'Farm', 'Farmer', 'Region', 'Weight(Lbs)', 'Variety',
       'Process'],
      dtype='object')


The columns in coffee23 are: 
Index(['Rank', 'Score', 'Farmer', 'Farm', 'Region', 'Estimated Weight (kg)',
       'Process', 'Variety'],
      dtype='object')




In [13]:
# Updating the auctions group data
auctions = auction21, auction22, auction23

# Checking columns names in auctions group
for a, t in zip (auctions, texts):
    cols = a.columns
    print(f"The columns in {t} are: \n{cols}\n\n")

The columns in auction21 are: 
Index(['Rank', 'Farm', 'Score', 'Weight (Lbs)', 'High Bid', 'Total Value',
       'Buyer'],
      dtype='object')


The columns in auction22 are: 
Index(['Rank', 'Farm', 'Score', 'Weight (Lbs)', 'High Bid', 'Total Value',
       'Buyer'],
      dtype='object')


The columns in auction23 are: 
Index(['Rank', 'Score', 'Farm', 'Weight (Lbs)', 'High Bid', 'Total Value',
       'Buyer'],
      dtype='object')




Now the data column names throughout the data are the same, although the coffees group still has a different column order, but it is not a problem in this case.

In [14]:
coffee21.head(3)

Rank  Score           Farm                   Farmer      Region  \
0     1  89.28  Pantan Musara           Dilen Ali Gogo        Aceh   
1     2  89.04       Ibun Ita               Ita Rosita  Jawa Barat   
2     3  88.89  Pantan Musara  Roberto Bagus Syahputra        Aceh   

   Weight (Kg)             Variety  Process  
0          210  Ateng, Gayo 1, P88    Honey  
1          186        Sigararutang  Natural  
2          262  Ateng, Gayo 1, P88   Washed

In [15]:
auction21.head(3)

Rank           Farm  Score  Weight (Lbs) High Bid Total Value  \
0    1  Pantan Musara  89.28        462.97   $80.00  $37,037.60   
1    2       Ibun Ita  89.04        410.06   $65.20  $26,735.91   
2    3  Pantan Musara  88.89        577.61   $69.10  $39,912.85   

                        Buyer  
0  Wataru for YAMATOYA COFFEE  
1  Proud Mary Coffee Roasters  
2   Terarosa (Haksan Co. Ltd)

## Null values handling.

In [16]:
# Create a list of DataFrames
all = [coffee21, auction21, coffee22, auction22, coffee23, auction23]

# Create a list of strings that represent the DataFrames
all_strings = ['coffee21', 'auction21', 'coffee22', 'auction22', 'coffee23', 'auction23']

# Checking null values
for i, n in zip(all, all_strings):
    null_status = i.isnull().values.any()
    if null_status:
        print(n + " has null values")
    else:
        print(n + " doesn't have null values")

coffee21 doesn't have null values
auction21 has null values
coffee22 doesn't have null values
auction22 has null values
coffee23 doesn't have null values
auction23 has null values


Checking a DataFrame that contains null value.

In [17]:
auction21.tail(3)

Rank      Farm  Score  Weight (Lbs)   High Bid  Total Value  \
24       25  Wajamala  87.16        478.40     $18.20    $8,706.88   
25       26   HT Farm  87.10        396.83     $17.60    $6,984.21   
26  Totals:       NaN    NaN      12195.97  $29.55/lb  $360,342.58   

                 Buyer  
24  Wataru & Co., Ltd.  
25          Blooms Ltd  
26                 NaN

In [18]:
# Check the null value occurrence in 2021 data
auction21.isnull().any(axis=1)

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26     True
dtype: bool

In [19]:
# Check the null value occurrence in 2022 data
auction22.isnull().any(axis=1)

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23     True
dtype: bool

In [20]:
# Check the null value occurrence in 2023 data
auction23.isnull().any(axis=1)

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25     True
dtype: bool

The null values is found in the last row of the 'auctions' group.

In [21]:
# Create a list of DataFrames
auctions = [auction21, auction22, auction23]

# Drop the last row of each dataframe in the 'auctions' group
for i in range(len(auctions)):
    last_row = len(auctions[i]) - 1
    auctions[i] = auctions[i].drop(auctions[i].index[last_row], inplace = True)

Recheck the null values presence.

In [22]:
# Rechecking null values
for i, n in zip(all, all_strings):
    null_status = i.isnull().values.any()
    if null_status:
        print(n + " has null values")
    else:
        print(n + " doesn't have null values")

coffee21 doesn't have null values
auction21 doesn't have null values
coffee22 doesn't have null values
auction22 doesn't have null values
coffee23 doesn't have null values
auction23 doesn't have null values


## Columns "High Bid" and "Total Value" handling.

Let's check first the data types from each columns in the auctions group.

In [23]:
# Create a list of DataFrames
auctions = [auction21, auction22, auction23]
texts = ['auction21', 'auction22', 'auction23']

# Checking data types in each column in auctions group
for a, t in zip (auctions, texts):
    types = a.dtypes
    print(f"The data types of each column in {t} are: \n{types}\n\n")

The data types of each column in auction21 are: 
Rank             object
Farm             object
Score           float64
Weight (Lbs)    float64
High Bid         object
Total Value      object
Buyer            object
dtype: object


The data types of each column in auction22 are: 
Rank             object
Farm             object
Score           float64
Weight (Lbs)    float64
High Bid         object
Total Value      object
Buyer            object
dtype: object


The data types of each column in auction23 are: 
Rank             object
Score           float64
Farm             object
Weight (Lbs)     object
High Bid         object
Total Value      object
Buyer            object
dtype: object




### Convert the data type of column "Weight (Lbs)" in 2023 data

In [24]:
auction23["Weight (Lbs)"] = auction23["Weight (Lbs)"].astype(float)

### Remove the "$" symbol in the columns "High Bid" and "Total Value"

In [25]:
# Check the value in the columns "High Bid" and "Total Value"
for a, t in zip (auctions, texts):
    head_data = a[["Weight (Lbs)", "High Bid", "Total Value"]].head(3)
    print(f"The top 3 data in columns 'Weight (Lbs)', 'High Bid' and 'Total Value' in {t} are: \n{head_data}\n\n")

The top 3 data in columns 'Weight (Lbs)', 'High Bid' and 'Total Value' in auction21 are: 
   Weight (Lbs) High Bid Total Value
0        462.97   $80.00  $37,037.60
1        410.06   $65.20  $26,735.91
2        577.61   $69.10  $39,912.85


The top 3 data in columns 'Weight (Lbs)', 'High Bid' and 'Total Value' in auction22 are: 
   Weight (Lbs) High Bid  Total Value
0        429.90  $ 40.20  $ 17,281.98
1        396.83  $ 40.20  $ 15,952.57
2        595.25  $ 20.20  $ 12,024.05


The top 3 data in columns 'Weight (Lbs)', 'High Bid' and 'Total Value' in auction23 are: 
   Weight (Lbs) High Bid Total Value
0        396.83   $45.50  $18,055.77
1        391.19   $43.60  $17,055.88
2        330.69   $30.00   $9,920.70




The data in columns "High Bid" and "Total Value" are considered strings because they contain the '$' symbol; thus, we need to remove the symbol and convert them into floats.

In [26]:
import re

# Function to remove the "$" symbol
def remove_dollar(text):
    return re.sub(r'\$','',text).strip()

In [27]:
# Apply the function to remove the "$" symbol
dfs = [auction21, auction22, auction23]

for df in dfs:
    df["High Bid"] = df["High Bid"].apply(remove_dollar)
    df["Total Value"] = df["Total Value"].apply(remove_dollar)
    df["High Bid"] = df["High Bid"].apply(remove_dollar)
    df["Total Value"] = df["Total Value"].apply(remove_dollar)


### Convert the data in column "High Bid" into floats

The value in the column "High Bid" is meant to be a float, but it is recognized as a string instead.

There is one piece of data that is inconsistent; it just contains one decimal number instead of two like the rest, so we need to alter it first.

In [28]:
# Locate the data
auction22.iloc[18]['High Bid']

'13.2'

In [29]:
# Change the data value
auction22.at[18, 'High Bid'] = '13.20'

In [30]:
# Recheck the data in auction22 column "High Bid"
auction23['High Bid']

0     45.50
1     43.60
2     30.00
3     30.50
4     26.25
5     27.30
6     10.00
7      8.60
8     10.37
9     15.80
10     9.30
11    11.80
12     9.00
13     7.00
14     8.60
15    10.20
16    11.11
17    10.90
18    11.10
19    10.20
20    10.20
21    13.60
22     6.25
23    16.40
24     8.45
Name: High Bid, dtype: object

Now we can convert the strings into floats

In [31]:
for df in dfs:
    df["High Bid"] = df["High Bid"].astype(float) # Convert the strings data into floats

### Convert the data in column "Total Value" into floats

Because the string format from the raw data is not compatible with being directly converted into floats, we need to try another approach.

First, we clear the data contained in the column, resulting in an empty column, then refill the data with the result of multiplying the data in columns "Weight (Lbs)" and "High Bid".

In [32]:
for df in dfs:
    df["Total Value"] = '' # Clear the data, resulting in an empty column
    df["Total Value"] = df["High Bid"] * df["Weight (Lbs)" ]# Fill the empty column

Now let's recheck the result after data conversion.

In [33]:
# Recheck data types in each column in auctions group
for a, t in zip (auctions, texts):
    types = a.dtypes
    print(f"The data types of each column in {t} are: \n{types}\n\n")

The data types of each column in auction21 are: 
Rank             object
Farm             object
Score           float64
Weight (Lbs)    float64
High Bid        float64
Total Value     float64
Buyer            object
dtype: object


The data types of each column in auction22 are: 
Rank             object
Farm             object
Score           float64
Weight (Lbs)    float64
High Bid        float64
Total Value     float64
Buyer            object
dtype: object


The data types of each column in auction23 are: 
Rank             object
Score           float64
Farm             object
Weight (Lbs)    float64
High Bid        float64
Total Value     float64
Buyer            object
dtype: object




In [34]:
# Recheck the value in the columns "High Bid" and "Total Value"
for a, t in zip (auctions, texts):
    head_data = a[["High Bid", "Total Value"]].head(2)
    print(f"The data values in columns 'High Bid' and 'Total Value' in {t} are: \n{head_data}\n\n")

The data values in columns 'High Bid' and 'Total Value' in auction21 are: 
   High Bid  Total Value
0      80.0    37037.600
1      65.2    26735.912


The data values in columns 'High Bid' and 'Total Value' in auction22 are: 
   High Bid  Total Value
0      40.2    17281.980
1      40.2    15952.566


The data values in columns 'High Bid' and 'Total Value' in auction23 are: 
   High Bid  Total Value
0      45.5    18055.765
1      43.6    17055.884




Now the data type of columns "High Bid" and "Total Value" is already converted into floats; however, the value in column "Total Value" has three decimal places, so we need to convert them to two decimal places to make them consistent with the value in column "High Bid".

In [35]:
pd.options.display.float_format = '{:.2f}'.format

In [36]:
# Round the decimal to two places
# for df in dfs:
#     df["Total Value"] = df["Total Value"].round(2)

### Add the "(US$)" note to the columns related to price 

Last, we need to alter the columns "High Bid" and "Total Value" with "(US$)" note to describe the unit.

In [37]:
# Define the dictionary for column names conversion
cols_conv = {"High Bid":  "High Bid (US$)", "Total Value": "Total Value (US$)"}

In [38]:
# Convert the column names in auctions group
for df in dfs:
    df.rename(columns = cols_conv, inplace = True)

In [39]:
# Recheck the value in the columns "High Bid" and "Total Value"
for a, t in zip (auctions, texts):
    head_data = a[["High Bid (US$)", "Total Value (US$)"]].head(3)
    print(f"The data values in columns 'High Bid' and 'Total Value' in {t} are: \n{head_data}\n\n")

The data values in columns 'High Bid' and 'Total Value' in auction21 are: 
   High Bid (US$)  Total Value (US$)
0           80.00           37037.60
1           65.20           26735.91
2           69.10           39912.85


The data values in columns 'High Bid' and 'Total Value' in auction22 are: 
   High Bid (US$)  Total Value (US$)
0           40.20           17281.98
1           40.20           15952.57
2           20.20           12024.05


The data values in columns 'High Bid' and 'Total Value' in auction23 are: 
   High Bid (US$)  Total Value (US$)
0           45.50           18055.76
1           43.60           17055.88
2           30.00            9920.70




In [40]:
auction21.head(3)

Rank           Farm  Score  Weight (Lbs)  High Bid (US$)  Total Value (US$)  \
0    1  Pantan Musara  89.28        462.97           80.00           37037.60   
1    2       Ibun Ita  89.04        410.06           65.20           26735.91   
2    3  Pantan Musara  88.89        577.61           69.10           39912.85   

                        Buyer  
0  Wataru for YAMATOYA COFFEE  
1  Proud Mary Coffee Roasters  
2   Terarosa (Haksan Co. Ltd)

## Columns "Weight" handling.

The unit of measurement in the columns "Weight" in the 'coffees' group are different from the unit in the 'auctions' group.

In [41]:
# Check the column name related to "Weight"
for c, s, a, t in zip (coffees, strings, auctions, texts):
    col_name = c.columns[5]
    print(f"The column name related to 'Weight' {s} is: \n{col_name}\n")

The column name related to 'Weight' coffee21 is: 
Weight (Kg)

The column name related to 'Weight' coffee22 is: 
Weight(Lbs)

The column name related to 'Weight' coffee23 is: 
Estimated Weight (kg)



In [42]:
auction22.head(2)

Rank                         Farm  Score  Weight (Lbs)  High Bid (US$)  \
0   1a  Koperasi Ribang Gayo Musara  90.59        429.90           40.20   
1   1b  Koperasi Ribang Gayo Musara  90.59        396.83           40.20   

   Total Value (US$)                                              Buyer  
0           17281.98                             Sulalat Coffee Trading  
1           15952.57  Coffee Beyond Borders for Uminomuko Coffee, Xtrat

Because of the coffees are bidded in the lbs unit in the auction, I decide to drop the columns related "Weight" in the `coffees` group, the quantity of coffee lots are specified in the 'auctions' group.

In [43]:
# Dropping columns related to "Weight" in the 'coffees' group.
coffee21.drop(['Weight (Kg)'], axis = 1, inplace = True)
coffee22.drop(['Weight(Lbs)'], axis = 1, inplace = True)
coffee23.drop(['Estimated Weight (kg)'], axis = 1, inplace = True)

## Column "Region" handling.

The column "Region" in the 'coffees' group refers to the the province where the farm is located. 

The value in the “Region” column for the year 2023 is written in English, while the previous 2 years are written in Indonesian.

 I prefer to alter them into Indonesian to make it consistent.



In [44]:
# Check the unique values in columns "Region"
pd.concat(coffees)['Region'].unique()

array(['Aceh', 'Jawa Barat', 'Jawa Timur', 'Jambi', 'Sulawesi Selatan',
       'Bali', 'NTT', 'East Java', 'East Nusa Tenggara', 'South Sulawesi',
       'West Java', 'Jambi, Central Sumatra'], dtype=object)

In [45]:
# Checking the unique values of "Region" columns in coffees group
for c, s in zip (coffees, strings):
    unique = c['Region'].unique()
    print(f"The unique values of column 'Region' in {s} are: \n{unique}\n\n")

The unique values of column 'Region' in coffee21 are: 
['Aceh' 'Jawa Barat' 'Jawa Timur' 'Jambi' 'Sulawesi Selatan' 'Bali' 'NTT']


The unique values of column 'Region' in coffee22 are: 
['Aceh' 'Jawa Barat' 'Jambi' 'Jawa Timur']


The unique values of column 'Region' in coffee23 are: 
['East Java' 'East Nusa Tenggara' 'South Sulawesi' 'Aceh' 'West Java'
 'Bali' 'Jambi, Central Sumatra']




From the list of strings above, we can convert some of them as follows:
- 'Jambi, Central Sumatra' to 'Jambi'
- 'West Java' to 'Jawa Barat'
- 'East Java' to 'Jawa Timur'
- 'South Sulawesi' to 'Sulawesi Selatan'
- 'East Nusa Tenggara' to 'Nusa Tenggara Timur'


In [46]:
# Define the dictionary for column names replacements
region = {'Jambi, Central Sumatra': 'Jambi', 'West Java': 'Jawa Barat', 'East Java': 'Jawa Timur',\
    'South Sulawesi': 'Sulawesi Selatan', 'East Nusa Tenggara': 'NTT'}

In [47]:
coffee23.replace(region, inplace = True)

## Column "Variety" handling.

This chapter is the most challenging work in the Data Wrangling work.

Let's check the unique values of column "Variety" throughout the data in coffees group.

In [48]:
# Checking the unique values of "Variety" columns in coffees group
for c, s in zip (coffees, strings):
    unique = c['Variety'].unique()
    print(f"The unique values of column 'Variety' in {s} are: \n{unique}\n\n")

The unique values of column 'Variety' in coffee21 are: 
['Ateng, Gayo 1, P88' 'Sigararutang' 'USDA, Colombia Brazil'
 'Sigararutang, Kartika, S-795' 'Sigararutang, S-795, Andungsari'
 'Abyssinia, P88, Ateng, Typica' 'Ateng, Sigararutang, S-795, Andungsari'
 'Typica' 'Cobra & Typica' 'S-795, USDA, Catuai'
 'Lini S 795, Catimor, Sigararutang' 'Typica& P-88' 'Ateng & Sigararutang'
 'S-795 & Cobra' 'S-795' 'Sigararutang & S-795'
 'S-795, Sigararutang, Yellow Bourbon' 'Andungsari & S-795' 'Ateng & P88'
 'Ateng' 'Andungsari & Sigararutang']


The unique values of column 'Variety' in coffee22 are: 
['Ateng 30%, P-88 70%' 'Sigararutang 70%, Typica 30%'
 'Sigararutang 16%, S-795 50%, P-88 17%, Bourbon 17%'
 'Ateng 50%, Bourbon 20%, P-88 30%' 'Ateng 65%, P-88 35%'
 'Ateng 40%, Typica 20%, Tim-Tim 40%'
 'Ateng 40%, Typica 20%, Caturra 20%, P-88 20%'
 'Ateng 30%, Typica 40%, Yang Lain 30%' 'Sigararutang 70%, Andungsari 30%'
 'USDA 80%, Kartika 20%' 'Ateng 25%, Typica 20%, P-88 30%, Yang Lain 25%'


In [49]:
# Check the value in the columns "Variety"
for c, s in zip (coffees, strings):
    head_data = c[["Variety"]].head(5)
    print(f"The top 5 data in columns 'Variety' in {s} are: \n{head_data}\n\n")

The top 5 data in columns 'Variety' in coffee21 are: 
                        Variety
0            Ateng, Gayo 1, P88
1                  Sigararutang
2            Ateng, Gayo 1, P88
3         USDA, Colombia Brazil
4  Sigararutang, Kartika, S-795


The top 5 data in columns 'Variety' in coffee22 are: 
                                             Variety
0                                Ateng 30%, P-88 70%
1                                Ateng 30%, P-88 70%
2                       Sigararutang 70%, Typica 30%
3  Sigararutang 16%, S-795 50%, P-88 17%, Bourbon...
4                   Ateng 50%, Bourbon 20%, P-88 30%


The top 5 data in columns 'Variety' in coffee23 are: 
                         Variety
0  USDA 762 (75%); Kartika (25%)
1  USDA 762 (75%); Kartika (25%)
2       S-795 (95%); Typica (5%)
3       S-795 (95%); Typica (5%)
4              Lini S-795 (100%)




We can see that the values are inconsistent.

In 2021 data, there is no percentage of variety written, and the separator is the "," character in the multi-varieties coffee lot.
In 2022 data, the percentage of variety was written, and the separator is the "," character in the multi-varieties coffee lot.
In 2023 data, the percentage of variety was written, and the separator is the ";" character in the multi-varieties coffee lot.

The objective of this approach is to divide the value in column "Variety" into columns each containing one variety.

Let's make the user-defined function first.

In [50]:
def split_varieties(df, col='Variety'):
    """This function splits the string into a particular column in a DataFrame
    using the regular expression to define the separator and 
    distributes the results into four new columns.
    The original column is dropped.
    
    Args:
        df (pd.DataFrame): define the DataFrame
        col (str): define the column
    
    Return:
        df (pd.DataFrame): the updated DataFrame with expanded columns that is resulted
                           string division, while the original column is dropped.
    """
    
    a = df[col].str.split(r',\s|&|\s&\s', expand=True)
    df = df.join(a.rename(columns={0:'Variety_1', 1:'Variety_2', 2:'Variety_3', 3:'Variety_4'}))
    df = df.drop(col, axis=1)
    
    return df

### 2021

In [51]:
# Apply the function to the 2021 data
coffee21 = split_varieties(coffee21)

In [52]:
# Check the column names after the function applied
coffee21.columns

Index(['Rank', 'Score', 'Farm', 'Farmer', 'Region', 'Process', 'Variety_1',
       'Variety_2', 'Variety_3', 'Variety_4'],
      dtype='object')

In [53]:
vars = ['Variety_1', 'Variety_2', 'Variety_3', 'Variety_4']
c = coffee21
str = 'coffee21'

# Check the unique values in each new "Variety" columns in 2021
for v in vars:
    unique = c[v].unique()
    print(f"The unique values of column {v} in {str} are: \n{unique}\n\n")

The unique values of column Variety_1 in coffee21 are: 
['Ateng' 'Sigararutang' 'USDA' 'Abyssinia' 'Typica' 'Cobra' 'S-795'
 'Lini S 795' 'Andungsari']


The unique values of column Variety_2 in coffee21 are: 
['Gayo 1' None 'Colombia Brazil' 'Kartika' 'S-795' 'P88' 'Sigararutang'
 'Typica' 'USDA' 'Catimor' ' P-88' 'Cobra']


The unique values of column Variety_3 in coffee21 are: 
['P88' None 'S-795' 'Andungsari' 'Ateng' 'Catuai' 'Sigararutang'
 'Yellow Bourbon']


The unique values of column Variety_4 in coffee21 are: 
[None 'Typica' 'Andungsari']




In [54]:
# Check the unique values summary from entire "Variety" columns
pd.unique(coffee21[['Variety_1', 'Variety_2', 'Variety_3', 'Variety_4']].values.ravel())

array(['Ateng', 'Gayo 1', 'P88', None, 'Sigararutang', 'USDA',
       'Colombia Brazil', 'Kartika', 'S-795', 'Andungsari', 'Abyssinia',
       'Typica', 'Cobra', 'Catuai', 'Lini S 795', 'Catimor', ' P-88',
       'Yellow Bourbon'], dtype=object)

There is one value in column "Variety_2" that still has a space character at the beginning of the value; we will alter it manually.

In [55]:
# Define the dictionary for data in column 'Variety_2' replacements
replace = {' P-88': 'P-88'}

# Modify the data in column 'Variety_2'
coffee21.replace(replace, inplace = True)

In [56]:
coffee21

Rank  Score                           Farm                   Farmer  \
0      1  89.28                  Pantan Musara           Dilen Ali Gogo   
1      2  89.04                       Ibun Ita               Ita Rosita   
2      3  88.89                  Pantan Musara  Roberto Bagus Syahputra   
3      4  88.75                   Ijen Lestari           Dandy Darmawan   
4      5  88.58                      Ibun Yudi                     Yudi   
5      6  88.49     Koperasi Koerintji Barokah                  Triyono   
6      7  88.46                   Avatara Gayo               Drs Hamdan   
7      8  88.30                       Kamojang             Ahmad Vansyu   
8      9  88.25                         Topidi              Daeng Halim   
9     10  88.15  Lereng Gunung Argopuro Krucil         Dinul Haq Sabyli   
10    11  88.14                         Topidi         Daeng Balengkang   
11    12  88.05          PT Sulotco Jaya Abadi         Samuel Karundeng   
12    13  87.88        Malabar Mountain Coffee           Muhammad Irwan   
13    14  87.76            Baru Dadap Petak 46              Saeful Hadi   
14    15  87.72                    Ketapanrame                  Wardoyo   
15    16  87.71                      Sukaramai                 Sabarwin   
16    17  87.69                      Batukaang            I Wayan Parum   
17    18  87.68                    Angin-angin                Indo Pole   
18    19  87.62              Gravfarm Sukaratu       Gravfarm Indonesia   
19    20  87.62                     Papandayan          Enung Sumartini   
20    21  87.53                LPE Kelola Agro             Setra Yuhana   
21    22  87.22                       Sarongge                  Santoso   
22    23  87.21                  Pantan Musara                   Zakiah   
23    24  87.21                     Bukit Sama            Dasimah Hakim   
24    25  87.16                       Wajamala           Marselina Walu   
25    26  87.10                        HT Farm                Mukhlis M   

              Region   Process     Variety_1        Variety_2       Variety_3  \
0               Aceh     Honey         Ateng           Gayo 1             P88   
1         Jawa Barat   Natural  Sigararutang             None            None   
2               Aceh    Washed         Ateng           Gayo 1             P88   
3         Jawa Timur   Natural          USDA  Colombia Brazil            None   
4         Jawa Barat   Natural  Sigararutang          Kartika           S-795   
5              Jambi     Honey  Sigararutang            S-795      Andungsari   
6               Aceh   Natural     Abyssinia              P88           Ateng   
7         Jawa Barat    Washed         Ateng     Sigararutang           S-795   
8   Sulawesi Selatan    Washed        Typica             None            None   
9         Jawa Timur   Natural         Cobra           Typica            None   
10  Sulawesi Selatan   Natural        Typica             None            None   
11  Sulawesi Selatan    Washed         S-795             USDA          Catuai   
12        Jawa Barat   Natural    Lini S 795          Catimor    Sigararutang   
13        Jawa Barat    Washed        Typica             None            None   
14        Jawa Timur   Natural        Typica             P-88            None   
15              Aceh  Wet Hull         Ateng     Sigararutang            None   
16              Bali   Natural         S-795            Cobra            None   
17  Sulawesi Selatan   Natural         S-795             None            None   
18        Jawa Barat    Washed  Sigararutang            S-795            None   
19        Jawa Barat   Natural         S-795     Sigararutang  Yellow Bourbon   
20        Jawa Barat   Natural    Andungsari            S-795            None   
21        Jawa Barat    Washed  Sigararutang            S-795      Andungsari   
22              Aceh   Natural         Ateng              P88            None   
23              Aceh  Wet Hull    

### 2022

In the years 2022 and 2023, the data in column 'Variety' consists of its percentage for the multi-varieties lot.

If we observe the detail, some of them are listed not in order from the higher percentage to the lower, so we need to revise it.

Also, there is one value listed in the Indonesian language, "Yang Lain", we will convert it to "Other (unspecified)".

In [57]:
# Check the unique values in "Variety" column in 2022 data 
coffee22['Variety'].unique()

array(['Ateng 30%, P-88 70%', 'Sigararutang 70%, Typica 30%',
       'Sigararutang 16%, S-795 50%, P-88 17%, Bourbon 17%',
       'Ateng 50%, Bourbon 20%, P-88 30%', 'Ateng 65%, P-88 35%',
       'Ateng 40%, Typica 20%, Tim-Tim 40%',
       'Ateng 40%, Typica 20%, Caturra 20%, P-88 20%',
       'Ateng 30%, Typica 40%, Yang Lain 30%',
       'Sigararutang 70%, Andungsari 30%', 'USDA 80%, Kartika 20%',
       'Ateng 25%, Typica 20%, P-88 30%, Yang Lain 25%',
       'Ateng 40%, P-88 30%, Gayo 1 30%',
       'Sigararutang 60%, Andungsari 40%', 'Ateng 100%',
       'Sigararutang 40%, S-795 50%, Andungsari 10%', 'Ateng'],
      dtype=object)

Sort the percents in the "Variety" column in descending order.

In [58]:
# Define the dictionary for data in column 'Variety' replacements
dict_coffee22 = {'Ateng 30%, P-88 70%': 'Ateng 70%, P-88 30%',\
        'Sigararutang 16%, S-795 50%, P-88 17%, Bourbon 17%': 'S-795 50%, P-88 17%, Bourbon 17%, Sigararutang 16%',\
        'Ateng 50%, Bourbon 20%, P-88 30%': 'Ateng 50%, P-88 30%, Bourbon 20%',\
        'Ateng 40%, Typica 20%, Tim-Tim 40%': 'Ateng 40%, Tim-Tim 40%, Typica 20%',\
        'Ateng 30%, Typica 40%, Yang Lain 30%': 'Typica 40%, Ateng 30%, Other (unspecified) 30%',\
        'Ateng 25%, Typica 20%, P-88 30%, Yang Lain 25%': 'P-88 30%, Ateng 25%, Other (unspecified) 25%, Typica 20%',\
        'Sigararutang 40%, S-795 50%, Andungsari 10%': 'S-795 50%, Sigararutang 40%, Andungsari 10%'}

In [59]:
# Modify the data in column 'Variety'
coffee22.replace(dict_coffee22, inplace = True)

We have the data in column "Variety" sorted, so the biggest percentage is mentioned first.

To make it easier to apply the same function "split_varieties()" later, we will drop the character "%" from the data, so the separator is the character ",".

In [60]:
# Function to remove 'any number %' from the data
def replace_percentage(text):
    return re.sub(r' \d+%','',text)

In [61]:
# Apply the function
coffee22['Variety'] = coffee22['Variety'].apply(replace_percentage)
coffee22.head(3)

Rank  Score                         Farm                   Farmer Region  \
0   1A  90.59  Koperasi Ribang Gayo Musara          Bahagia Ginting   Aceh   
1   1B  90.59  Koperasi Ribang Gayo Musara          Bahagia Ginting   Aceh   
2    2  89.85      Pegasing Takengon Utara  PT. Biji Kopi Internusa   Aceh   

                Variety  Process  
0           Ateng, P-88  Natural  
1           Ateng, P-88  Natural  
2  Sigararutang, Typica  Natural

The data now contain the character "," as separator, thus will fit the "split_varieties()" function.

In [62]:
# Apply the function to the 2022 data
coffee22 = split_varieties(coffee22)

In [63]:
# Check the column names after the function applied
coffee22.columns

Index(['Rank', 'Score', 'Farm', 'Farmer', 'Region', 'Process', 'Variety_1',
       'Variety_2', 'Variety_3', 'Variety_4'],
      dtype='object')

In [64]:
vars = ['Variety_1', 'Variety_2', 'Variety_3', 'Variety_4']
c = coffee22
str = 'coffee22'

# Check the unique values in each new "Variety" columns in 2022
for v in vars:
    unique = c[v].unique()
    print(f"The unique values of column {v} in {str} are: \n{unique}\n\n")

The unique values of column Variety_1 in coffee22 are: 
['Ateng' 'Sigararutang' 'S-795' 'Typica' 'USDA' 'P-88']


The unique values of column Variety_2 in coffee22 are: 
['P-88' 'Typica' 'Tim-Tim' 'Ateng' 'Andungsari' 'Kartika' None
 'Sigararutang']


The unique values of column Variety_3 in coffee22 are: 
[None 'Bourbon' 'Typica' 'Caturra' 'Other (unspecified)' 'Gayo 1'
 'Andungsari']


The unique values of column Variety_4 in coffee22 are: 
[None 'Sigararutang' 'P-88' 'Typica']




In [65]:
# Check the unique values summary from entire "Variety" columns
pd.unique(coffee22[['Variety_1', 'Variety_2', 'Variety_3', 'Variety_4']].values.ravel())

array(['Ateng', 'P-88', None, 'Sigararutang', 'Typica', 'S-795',
       'Bourbon', 'Tim-Tim', 'Caturra', 'Other (unspecified)',
       'Andungsari', 'USDA', 'Kartika', 'Gayo 1'], dtype=object)

In [66]:
coffee22

Rank  Score                          Farm                   Farmer  \
0    1A  90.59   Koperasi Ribang Gayo Musara          Bahagia Ginting   
1    1B  90.59   Koperasi Ribang Gayo Musara          Bahagia Ginting   
2     2  89.85       Pegasing Takengon Utara  PT. Biji Kopi Internusa   
3     3  89.59            Frinsa Weninggalih  Yasmin Nabila Salsabila   
4     4  89.18   Koperasi Ribang Gayo Musara                   Zakiah   
5     5  89.15   Koperasi Ribang Gayo Musara           Dimas Apriandi   
6     6  88.91  Panangan Mata Takengon Utara                Christian   
7     7  88.68   Koperasi Ribang Gayo Musara            Asman Arianto   
8     8  88.50   Koperasi Ribang Gayo Musara           Dilen Ali Gogo   
9     9  88.09                   Terang Ulen              Fahman Yoga   
10   10  87.85     Ceding Ayu, Pantan Musara               Drs Hamdan   
11   11  87.74                      Renggali            Hakem Murtada   
12   12  87.74       Koerintji Barokah Group       Abdul Aziz Thohari   
13   13  87.65                Koerintji Farm          Ahmad Abi Sodik   
14   14  87.62     Ceding Ayu, Pantan Musara                Eriyani N   
15   15  87.62                          Ijen          Dandy Dharmawan   
16   16  87.59                Avatara Coffee          Hendra Maulizar   
17   17  87.53   Koperasi Ribang Gayo Musara              Rudi Suwito   
18   18  87.47   Koperasi Ribang Gayo Musara  Roberto Bagus Syahputra   
19   19  87.35               Tunas Baru Farm                  Irwandi   
20   20  87.15            Dasimah bukit sama                    Yurda   
21   21  87.15                Jaya Tani Farm                  Triyono   
22   22  87.06   Koperasi Ribang Gayo Musara       Victor Nopriansyah   

        Region                    Process     Variety_1     Variety_2  \
0         Aceh                    Natural         Ateng          P-88   
1         Aceh                    Natural         Ateng          P-88   
2         Aceh                    Natural  Sigararutang        Typica   
3   Jawa Barat          Anaerobic Natural         S-795          P-88   
4         Aceh                    Natural         Ateng          P-88   
5         Aceh                     Washed         Ateng          P-88   
6         Aceh                      Honey  Sigararutang        Typica   
7         Aceh                      Honey         Ateng          P-88   
8         Aceh                      Honey         Ateng          P-88   
9         Aceh                 Wet Hulled         Ateng       Tim-Tim   
10        Aceh  Carbonic Maceration Honey         Ateng        Typica   
11        Aceh                 Wet Hulled        Typica         Ateng   
12       Jambi             Lactic Natural  Sigararutang    Andungsari   
13       Jambi                     Washed  Sigararutang    Andungsari   
14        Aceh          Anaerobic Natural         Ateng        Typica   
15  Jawa Timur          Anaerobic Natural          USDA       Kartika   
16        Aceh                     Washed          P-88         Ateng   
17        Aceh                     Washed         Ateng          P-88   
18        Aceh                     Washed         Ateng          P-88   
19       Jambi          Anaerobic Natural  Sigararutang    Andungsari   
20        Aceh                      Honey         Ateng          None   
21       Jambi          Anaerobic Natural         S-795  Sigararutang   
22        Aceh                     Washed         Ateng          None   

              Variety_3     Variety_4  
0                  None          None  
1                  None          None  
2                  None          None  
3               Bourbon  Sigararutang  
4               Bourbon          None  
5                  None          None  
6                  None          None  
7               Bourbon          None  
8               Bourbon          None  
9                Typica          None  
10              Caturra          P-88  
11  Other (unspec

### 2023

The 2023 data will be handled the same way as the 2022 data.

In [67]:
# Check the unique values in "Variety" column in 2023 data 
coffee23['Variety'].unique()

array(['USDA 762 (75%); Kartika (25%)', 'S-795 (95%); Typica (5%)',
       'Lini S-795 (100%)',
       'Ateng (40%); Typica (15%); P-88 (25%); yang lain (20%)',
       'Bourbon (20%); Typica (50%); Abyssinia (30%)',
       'Ateng (30%); Sigararutang (40%); S-795 (30%)',
       'Gayo 3 (50%), Gayo 1 (50%',
       'USDA (20%); S-795 (30%); Caturra (50%)', 'Ateng (100%)',
       'Ateng (20%); S-795 (80%)', 'Ateng (70%); P-88 (30%)',
       'Kopyol (100%)', 'Ateng (60%); Sigararutang (40%)',
       'Sigararutang (90%); Andungsari (10%)',
       'Ateng (35%); Sigararutang (30%); S-795 (35%)',
       'Sigararutang (100%)',
       'Ateng (50%); Sigararutang (30%); S-795 (15%); P-88 (5%)',
       'Ateng (50%); Typica (10%); P-88 (40%)', 'Ateng (90%)',
       'Bourbon (50%); Typica (50%)',
       'Sigararutang (75%); Andungsari (25%)', 'Red Bourbon (100%)'],
      dtype=object)

Sort the percents in the "Variety" column in descending order.

In [68]:
# Define the dictionary for data in column 'Variety' replacements
dict_coffee23 = {'Ateng (40%); Typica (15%); P-88 (25%); yang lain (20%)': 'Ateng (40%); P-88 (25%); Other (unspecified) (20%); Typica (15%)',\
        'Bourbon (20%); Typica (50%); Abyssinia (30%)': 'Typica (50%); Abyssinia (30%); Bourbon (20%)',\
        'Ateng (30%); Sigararutang (40%); S-795 (30%)': 'Sigararutang (40%); Ateng (30%); S-795 (30%)',\
        'USDA (20%); S-795 (30%); Caturra (50%)': 'Caturra (50%); S-795 (30%); USDA (20%)',\
        'Ateng (20%); S-795 (80%)': 'S-795 (80%); Ateng (20%)',\
        'Ateng (35%); Sigararutang (30%); S-795 (35%)': 'Ateng (35%); S-795 (35%); Sigararutang (30%)',\
        'Ateng (50%); Typica (10%); P-88 (40%)': 'Ateng (50%); P-88 (40%); Typica (10%)',\
        }

In [69]:
# Modify the data in column 'Variety'
coffee23.replace(dict_coffee23, inplace = True)

In [70]:
# Function to replace (any number%) with 'column_value'
def replace_percentage(text):
    return re.sub(r' .\d+%.| .\d+%','',text)

# Function to replace the character ";" with ","
def replace_semicolon(text):
    return re.sub(r';',',',text)

In [71]:
# Replace the '%' strings
coffee23['Variety'] = coffee23['Variety'].apply(replace_percentage)
coffee23['Variety'] = coffee23['Variety'].apply(replace_semicolon)

# Check the result
coffee23.head(3)

Rank  Score           Farmer          Farm      Region  \
0   1A  91.41  Dandy Dharmawan  Ijen Lestari  Jawa Timur   
1   1B  91.41  Dandy Dharmawan  Ijen Lestari  Jawa Timur   
2   2A  90.35   Marselina Walu      Wajamala         NTT   

                       Process            Variety  
0  Carbonic Maceration Natural  USDA 762, Kartika  
1  Carbonic Maceration Natural  USDA 762, Kartika  
2                   Dry Hulled      S-795, Typica

In [72]:
coffee23 = split_varieties(coffee23)

In [73]:
# Check the column names after the function applied
coffee23.columns

Index(['Rank', 'Score', 'Farmer', 'Farm', 'Region', 'Process', 'Variety_1',
       'Variety_2', 'Variety_3', 'Variety_4'],
      dtype='object')

In [74]:
vars = ['Variety_1', 'Variety_2', 'Variety_3', 'Variety_4']
c = coffee23
str = 'coffee23'

# Check the unique values in each new "Variety" columns in 2021
for v in vars:
    unique = c[v].unique()
    print(f"The unique values of column {v} in {str} are: \n{unique}\n\n")

The unique values of column Variety_1 in coffee23 are: 
['USDA 762' 'S-795' 'Lini S-795' 'Ateng' 'Typica' 'Sigararutang' 'Gayo 3'
 'Caturra' 'Kopyol' 'Bourbon' 'Red Bourbon']


The unique values of column Variety_2 in coffee23 are: 
['Kartika' 'Typica' None 'P-88' 'Abyssinia' 'Ateng' 'Gayo 1' 'S-795'
 'Sigararutang' 'Andungsari']


The unique values of column Variety_3 in coffee23 are: 
[None 'Other (unspecified)' 'Bourbon' 'S-795' 'USDA' 'Sigararutang'
 'Typica']


The unique values of column Variety_4 in coffee23 are: 
[None 'Typica' 'P-88']




In [75]:
# Check the unique values summary from entire "Variety" columns
pd.unique(coffee23[['Variety_1', 'Variety_2', 'Variety_3', 'Variety_4']].values.ravel())

array(['USDA 762', 'Kartika', None, 'S-795', 'Typica', 'Lini S-795',
       'Ateng', 'P-88', 'Other (unspecified)', 'Abyssinia', 'Bourbon',
       'Sigararutang', 'Gayo 3', 'Gayo 1', 'Caturra', 'USDA', 'Kopyol',
       'Andungsari', 'Red Bourbon'], dtype=object)

In [76]:
coffee23

Rank  Score                   Farmer  \
0    1A  91.41          Dandy Dharmawan   
1    1B  91.41          Dandy Dharmawan   
2    2A  90.35           Marselina Walu   
3    2B  90.35           Marselina Walu   
4    3A  90.06              Yolan Tirta   
5    3B  90.06              Yolan Tirta   
6     4  90.06          Hendra Maulizar   
7     5  89.47     Enzo Sauqi Hutabarat   
8     6  89.35          Hj. Eti Sumiati   
9     7  89.21         Syafjurian Putra   
10    8  89.15         Samuel Karundeng   
11    9  88.38                  Fauziah   
12   10  88.15            Diki Wahyudin   
13   11  88.06                   Zakiah   
14   12  88.03          I Wayan Suparta   
15   13  87.97            Nurdianingsih   
16   14  87.97      Eka Teguh Kurniawan   
17   15  87.82            Satrea Amambi   
18   16  87.65               Ita Rosita   
19   17  87.59  Yasmin Nabila Salsabila   
20   18  87.47            Zainal Arifin   
21   19  87.35                   Marjan   
22   20  87.35               Supriyanto   
23   21  87.32       Abdul Aziz Thohari   
24   22  87.24              Ferri Padli   

                                                 Farm            Region  \
0                                        Ijen Lestari        Jawa Timur   
1                                        Ijen Lestari        Jawa Timur   
2                                            Wajamala               NTT   
3                                            Wajamala               NTT   
4                                         Angin-angin  Sulawesi Selatan   
5                                         Angin-angin  Sulawesi Selatan   
6                                      Avarata Coffee              Aceh   
7                             Central Sumatera Coffee              Aceh   
8                  Wanoja Coffee Plot Hj. Eti Sumiati        Jawa Barat   
9                                       Atang Jungket              Aceh   
10  Rantekarua Coffee Plantation PT. Sulotco Jaya ...  Sulawesi Selatan   
11                         PT. De Gayo Agri Indonesia              Aceh   
12                                    Finca De Berman        Jawa Barat   
13                        Koperasi Ribang Gayo Musara              Aceh   
14                                      Pelage Coffee              Bali   
15                                 Bengkolan Dua Farm              Aceh   
16                                         Blok Emper        Jawa Barat   
17                   Wanoja Coffee Plot Satrea Amambi        Jawa Barat   
18                      Wanoja Coffee Plot Ita Rosita        Jawa Barat   
19                                        Weninggalih        Jawa Barat   
20                                  Kian Mandiri Farm              Aceh   
21                                       Siluk Coffee              Aceh   
22                        Suket Mountain/Gunung Suket        Jawa Timur   
23                                 Rukun Saudara Farm             Jambi   
24                                    Empus Reje Guru              Aceh   

                        Process     Variety_1     Variety_2  \
0   Carbonic Maceration Natural      USDA 762       Kartika   
1   Carbonic Maceration Natural      USDA 762       Kartika   
2                    Dry Hulled         S-795        Typica   
3                    Dry Hulled         S-795        Typica   
4                         Honey    Lini S-795          None   
5                         Honey    Lini S-795          None   
6                       Natural         Ateng          P-88   
7                       Natural        Typica     Abyssinia   
8                         Honey  Sigararutang         Ateng   
9                       Natural        Gayo 3        Gayo 1   
10                      Natural       Caturra         S-795   
11                      Natural         Ateng          None   
12              Honey Anaerobic         S-795         Ateng   
13                      Natural         Ateng      

### Final Inspections

The coffee varieties are already handled and separated into four new columns, each containing one unique variety.

Let's check the result on the DataFrames.

In [88]:
coffees = [coffee21, coffee22, coffee23]
vars = ['Variety_1', 'Variety_2', 'Variety_3', 'Variety_4']

# Combine DataFrames
combined_df = pd.concat(coffees, ignore_index=True)

# Select the columns
combined_vars = combined_df[vars]

# Get unique values
unique_vars = pd.unique(combined_vars.values.ravel())
unique_vars[pd.notnull(unique_vars)]


array(['Ateng', 'Gayo 1', 'P88', 'Sigararutang', 'USDA',
       'Colombia Brazil', 'Kartika', 'S-795', 'Andungsari', 'Abyssinia',
       'Typica', 'Cobra', 'Catuai', 'Lini S 795', 'Catimor', 'P-88',
       'Yellow Bourbon', 'Bourbon', 'Tim-Tim', 'Caturra',
       'Other (unspecified)', 'USDA 762', 'Lini S-795', 'Gayo 3',
       'Kopyol', 'Red Bourbon'], dtype=object)

There are some values that suspiciously have the same meaning/defintion, which are:
- 'P88' and 'P-88'
- 'S-795', 'Lini S 795', and 'Lini S-795'
- 'USDA' and 'USDA 762'

This case will be noted to be discussed with the domain expert later in the data validation step.

## Dataframe Merging

After the data are cleaned, we now need to merge those data.

First, the data from the coffees group is merged with the auctions group from the same year horizontally, with the duplicate columns being dropped to avoid confusion.

After we get the merged data from each year, we will add a new column indicating the year of competition.

Lastly, the merged data from each year will be combined into one dataset.

Here we define the function:

In [89]:
def COE_year(dfs: list, 
              y: str, 
              save_file: bool = True)-> pd.DataFrame:
    """
    Args:
        dfs (list): contains the DataFrames of the same year
        y (str): year of competition
        save_file (bool, optional): save the DataFrame as .csv format, default to True
    Returns:
        A merged DataFrame with no duplicate columns, inserted by a new column refer to year.
    """
    result = pd.concat(dfs, axis=1)
    result = result.loc[:,~result.columns.duplicated()]
    result.insert(0, 'Year', y)
    PATH = '/home/er_bim/Indonesia_COE/data/processed/cleaned/'
    
    if save_file:  
        result.to_csv(PATH+'data'+y+'.csv')
        
    return result

### 2021

In [90]:
# Apply the function for year 2021 data
COE_year([coffee21, auction21], '2021')

Year  Rank  Score                           Farm                   Farmer  \
0   2021     1  89.28                  Pantan Musara           Dilen Ali Gogo   
1   2021     2  89.04                       Ibun Ita               Ita Rosita   
2   2021     3  88.89                  Pantan Musara  Roberto Bagus Syahputra   
3   2021     4  88.75                   Ijen Lestari           Dandy Darmawan   
4   2021     5  88.58                      Ibun Yudi                     Yudi   
5   2021     6  88.49     Koperasi Koerintji Barokah                  Triyono   
6   2021     7  88.46                   Avatara Gayo               Drs Hamdan   
7   2021     8  88.30                       Kamojang             Ahmad Vansyu   
8   2021     9  88.25                         Topidi              Daeng Halim   
9   2021    10  88.15  Lereng Gunung Argopuro Krucil         Dinul Haq Sabyli   
10  2021    11  88.14                         Topidi         Daeng Balengkang   
11  2021    12  88.05          PT Sulotco Jaya Abadi         Samuel Karundeng   
12  2021    13  87.88        Malabar Mountain Coffee           Muhammad Irwan   
13  2021    14  87.76            Baru Dadap Petak 46              Saeful Hadi   
14  2021    15  87.72                    Ketapanrame                  Wardoyo   
15  2021    16  87.71                      Sukaramai                 Sabarwin   
16  2021    17  87.69                      Batukaang            I Wayan Parum   
17  2021    18  87.68                    Angin-angin                Indo Pole   
18  2021    19  87.62              Gravfarm Sukaratu       Gravfarm Indonesia   
19  2021    20  87.62                     Papandayan          Enung Sumartini   
20  2021    21  87.53                LPE Kelola Agro             Setra Yuhana   
21  2021    22  87.22                       Sarongge                  Santoso   
22  2021    23  87.21                  Pantan Musara                   Zakiah   
23  2021    24  87.21                     Bukit Sama            Dasimah Hakim   
24  2021    25  87.16                       Wajamala           Marselina Walu   
25  2021    26  87.10                        HT Farm                Mukhlis M   

              Region   Process     Variety_1        Variety_2       Variety_3  \
0               Aceh     Honey         Ateng           Gayo 1             P88   
1         Jawa Barat   Natural  Sigararutang             None            None   
2               Aceh    Washed         Ateng           Gayo 1             P88   
3         Jawa Timur   Natural          USDA  Colombia Brazil            None   
4         Jawa Barat   Natural  Sigararutang          Kartika           S-795   
5              Jambi     Honey  Sigararutang            S-795      Andungsari   
6               Aceh   Natural     Abyssinia              P88           Ateng   
7         Jawa Barat    Washed         Ateng     Sigararutang           S-795   
8   Sulawesi Selatan    Washed        Typica             None            None   
9         Jawa Timur   Natural         Cobra           Typica            None   
10  Sulawesi Selatan   Natural        Typica             None            None   
11  Sulawesi Selatan    Washed         S-795             USDA          Catuai   
12        Jawa Barat   Natural    Lini S 795          Catimor    Sigararutang   
13        Jawa Barat    Washed        Typica             None            None   
14        Jawa Timur   Natural        Typica             P-88            None   
15              Aceh  Wet Hull         Ateng     Sigararutang            None   
16              Bali   Natural         S-795            Cobra            None   
17  Sulawesi Selatan   Natural         S-795             None            None   
18        Jawa Barat    Washed  Sigararutang            S-795            None   
19        Jawa Barat   Natural         S-795     Sigararutang  Yellow Bourbon   
20        Jawa Barat   Natural    Andungsari            S-795            None   
21        Jawa Barat    Washed  Si

In [91]:
PATH = '/home/er_bim/Indonesia_COE/data/processed/cleaned/'

# Data reading inspection for year 2021 data
data21 = pd.read_csv(PATH + 'data2021.csv', index_col=0)
data21.head(3)

Year  Rank  Score           Farm                   Farmer      Region  \
0  2021     1  89.28  Pantan Musara           Dilen Ali Gogo        Aceh   
1  2021     2  89.04       Ibun Ita               Ita Rosita  Jawa Barat   
2  2021     3  88.89  Pantan Musara  Roberto Bagus Syahputra        Aceh   

   Process     Variety_1 Variety_2 Variety_3 Variety_4  Weight (Lbs)  \
0    Honey         Ateng    Gayo 1       P88       NaN        462.97   
1  Natural  Sigararutang       NaN       NaN       NaN        410.06   
2   Washed         Ateng    Gayo 1       P88       NaN        577.61   

   High Bid (US$)  Total Value (US$)                       Buyer  
0           80.00           37037.60  Wataru for YAMATOYA COFFEE  
1           65.20           26735.91  Proud Mary Coffee Roasters  
2           69.10           39912.85   Terarosa (Haksan Co. Ltd)

### 2022

In [92]:
# Apply the function for year 2022 data
COE_year([coffee22, auction22], '2022')

Year Rank  Score                          Farm                   Farmer  \
0   2022   1A  90.59   Koperasi Ribang Gayo Musara          Bahagia Ginting   
1   2022   1B  90.59   Koperasi Ribang Gayo Musara          Bahagia Ginting   
2   2022    2  89.85       Pegasing Takengon Utara  PT. Biji Kopi Internusa   
3   2022    3  89.59            Frinsa Weninggalih  Yasmin Nabila Salsabila   
4   2022    4  89.18   Koperasi Ribang Gayo Musara                   Zakiah   
5   2022    5  89.15   Koperasi Ribang Gayo Musara           Dimas Apriandi   
6   2022    6  88.91  Panangan Mata Takengon Utara                Christian   
7   2022    7  88.68   Koperasi Ribang Gayo Musara            Asman Arianto   
8   2022    8  88.50   Koperasi Ribang Gayo Musara           Dilen Ali Gogo   
9   2022    9  88.09                   Terang Ulen              Fahman Yoga   
10  2022   10  87.85     Ceding Ayu, Pantan Musara               Drs Hamdan   
11  2022   11  87.74                      Renggali            Hakem Murtada   
12  2022   12  87.74       Koerintji Barokah Group       Abdul Aziz Thohari   
13  2022   13  87.65                Koerintji Farm          Ahmad Abi Sodik   
14  2022   14  87.62     Ceding Ayu, Pantan Musara                Eriyani N   
15  2022   15  87.62                          Ijen          Dandy Dharmawan   
16  2022   16  87.59                Avatara Coffee          Hendra Maulizar   
17  2022   17  87.53   Koperasi Ribang Gayo Musara              Rudi Suwito   
18  2022   18  87.47   Koperasi Ribang Gayo Musara  Roberto Bagus Syahputra   
19  2022   19  87.35               Tunas Baru Farm                  Irwandi   
20  2022   20  87.15            Dasimah bukit sama                    Yurda   
21  2022   21  87.15                Jaya Tani Farm                  Triyono   
22  2022   22  87.06   Koperasi Ribang Gayo Musara       Victor Nopriansyah   

        Region                    Process     Variety_1     Variety_2  \
0         Aceh                    Natural         Ateng          P-88   
1         Aceh                    Natural         Ateng          P-88   
2         Aceh                    Natural  Sigararutang        Typica   
3   Jawa Barat          Anaerobic Natural         S-795          P-88   
4         Aceh                    Natural         Ateng          P-88   
5         Aceh                     Washed         Ateng          P-88   
6         Aceh                      Honey  Sigararutang        Typica   
7         Aceh                      Honey         Ateng          P-88   
8         Aceh                      Honey         Ateng          P-88   
9         Aceh                 Wet Hulled         Ateng       Tim-Tim   
10        Aceh  Carbonic Maceration Honey         Ateng        Typica   
11        Aceh                 Wet Hulled        Typica         Ateng   
12       Jambi             Lactic Natural  Sigararutang    Andungsari   
13       Jambi                     Washed  Sigararutang    Andungsari   
14        Aceh          Anaerobic Natural         Ateng        Typica   
15  Jawa Timur          Anaerobic Natural          USDA       Kartika   
16        Aceh                     Washed          P-88         Ateng   
17        Aceh                     Washed         Ateng          P-88   
18        Aceh                     Washed         Ateng          P-88   
19       Jambi          Anaerobic Natural  Sigararutang    Andungsari   
20        Aceh                      Honey         Ateng          None   
21       Jambi          Anaerobic Natural         S-795  Sigararutang   
22        Aceh                     Washed         Ateng          None   

              Variety_3     Variety_4  Weight (Lbs)  High Bid (US$)  \
0                  None          None        429.90           40.20   
1                  None          None        396.83           40.20   
2                  None          None        595.25           20.20   
3               Bourbon  Sigararutang        595.25           26.90   

In [93]:
# Data reading inspection for year 2022 data
data22 = pd.read_csv(PATH + 'data2022.csv', index_col=0)
data22.head(3)

Year Rank  Score                         Farm                   Farmer  \
0  2022   1A  90.59  Koperasi Ribang Gayo Musara          Bahagia Ginting   
1  2022   1B  90.59  Koperasi Ribang Gayo Musara          Bahagia Ginting   
2  2022    2  89.85      Pegasing Takengon Utara  PT. Biji Kopi Internusa   

  Region  Process     Variety_1 Variety_2 Variety_3 Variety_4  Weight (Lbs)  \
0   Aceh  Natural         Ateng      P-88       NaN       NaN        429.90   
1   Aceh  Natural         Ateng      P-88       NaN       NaN        396.83   
2   Aceh  Natural  Sigararutang    Typica       NaN       NaN        595.25   

   High Bid (US$)  Total Value (US$)  \
0           40.20           17281.98   
1           40.20           15952.57   
2           20.20           12024.05   

                                               Buyer  
0                             Sulalat Coffee Trading  
1  Coffee Beyond Borders for Uminomuko Coffee, Xtrat  
2                                     La Boheme Cafe

### 2023

In [94]:
# Apply the function for year 2023 data
COE_year([coffee23, auction23], '2023')

Year Rank  Score                   Farmer  \
0   2023   1A  91.41          Dandy Dharmawan   
1   2023   1B  91.41          Dandy Dharmawan   
2   2023   2A  90.35           Marselina Walu   
3   2023   2B  90.35           Marselina Walu   
4   2023   3A  90.06              Yolan Tirta   
5   2023   3B  90.06              Yolan Tirta   
6   2023    4  90.06          Hendra Maulizar   
7   2023    5  89.47     Enzo Sauqi Hutabarat   
8   2023    6  89.35          Hj. Eti Sumiati   
9   2023    7  89.21         Syafjurian Putra   
10  2023    8  89.15         Samuel Karundeng   
11  2023    9  88.38                  Fauziah   
12  2023   10  88.15            Diki Wahyudin   
13  2023   11  88.06                   Zakiah   
14  2023   12  88.03          I Wayan Suparta   
15  2023   13  87.97            Nurdianingsih   
16  2023   14  87.97      Eka Teguh Kurniawan   
17  2023   15  87.82            Satrea Amambi   
18  2023   16  87.65               Ita Rosita   
19  2023   17  87.59  Yasmin Nabila Salsabila   
20  2023   18  87.47            Zainal Arifin   
21  2023   19  87.35                   Marjan   
22  2023   20  87.35               Supriyanto   
23  2023   21  87.32       Abdul Aziz Thohari   
24  2023   22  87.24              Ferri Padli   

                                                 Farm            Region  \
0                                        Ijen Lestari        Jawa Timur   
1                                        Ijen Lestari        Jawa Timur   
2                                            Wajamala               NTT   
3                                            Wajamala               NTT   
4                                         Angin-angin  Sulawesi Selatan   
5                                         Angin-angin  Sulawesi Selatan   
6                                      Avarata Coffee              Aceh   
7                             Central Sumatera Coffee              Aceh   
8                  Wanoja Coffee Plot Hj. Eti Sumiati        Jawa Barat   
9                                       Atang Jungket              Aceh   
10  Rantekarua Coffee Plantation PT. Sulotco Jaya ...  Sulawesi Selatan   
11                         PT. De Gayo Agri Indonesia              Aceh   
12                                    Finca De Berman        Jawa Barat   
13                        Koperasi Ribang Gayo Musara              Aceh   
14                                      Pelage Coffee              Bali   
15                                 Bengkolan Dua Farm              Aceh   
16                                         Blok Emper        Jawa Barat   
17                   Wanoja Coffee Plot Satrea Amambi        Jawa Barat   
18                      Wanoja Coffee Plot Ita Rosita        Jawa Barat   
19                                        Weninggalih        Jawa Barat   
20                                  Kian Mandiri Farm              Aceh   
21                                       Siluk Coffee              Aceh   
22                        Suket Mountain/Gunung Suket        Jawa Timur   
23                                 Rukun Saudara Farm             Jambi   
24                                    Empus Reje Guru              Aceh   

                        Process     Variety_1     Variety_2  \
0   Carbonic Maceration Natural      USDA 762       Kartika   
1   Carbonic Maceration Natural      USDA 762       Kartika   
2                    Dry Hulled         S-795        Typica   
3                    Dry Hulled         S-795        Typica   
4                         Honey    Lini S-795          None   
5                         Honey    Lini S-795          None   
6                       Natural         Ateng          P-88   
7                       Natural        Typica     Abyssinia   
8                         Honey  Sigararutang         Ateng   
9                       Natural        Gayo 3        Gayo 1   
10                      Natural       Caturra         S-795   
11                    

In [95]:
# Data reading inspection for year 2023 data
data23 = pd.read_csv(PATH + 'data2023.csv', index_col=0)
data23.head(3)

Year Rank  Score           Farmer          Farm      Region  \
0  2023   1A  91.41  Dandy Dharmawan  Ijen Lestari  Jawa Timur   
1  2023   1B  91.41  Dandy Dharmawan  Ijen Lestari  Jawa Timur   
2  2023   2A  90.35   Marselina Walu      Wajamala         NTT   

                       Process Variety_1 Variety_2 Variety_3 Variety_4  \
0  Carbonic Maceration Natural  USDA 762   Kartika       NaN       NaN   
1  Carbonic Maceration Natural  USDA 762   Kartika       NaN       NaN   
2                   Dry Hulled     S-795    Typica       NaN       NaN   

   Weight (Lbs)  High Bid (US$)  Total Value (US$)  \
0        396.83           45.50           18055.76   
1        391.19           43.60           17055.88   
2        330.69           30.00            9920.70   

                                               Buyer  
0                      Orsir International Co., Ltd.  
1  SAVAYA Coffee Market – USA WAVY Coffee Roaster...  
2  Coffee Me Up (Korea), Slick Coffee Co., Coffee...

### Concatenate data from all years

In [96]:
# Concatanate the data
coe_indonesia = pd.concat([data21, data22, data23], ignore_index=True)
coe_indonesia

Year Rank  Score                         Farm                   Farmer  \
0   2021    1  89.28                Pantan Musara           Dilen Ali Gogo   
1   2021    2  89.04                     Ibun Ita               Ita Rosita   
2   2021    3  88.89                Pantan Musara  Roberto Bagus Syahputra   
3   2021    4  88.75                 Ijen Lestari           Dandy Darmawan   
4   2021    5  88.58                    Ibun Yudi                     Yudi   
..   ...  ...    ...                          ...                      ...   
69  2023   18  87.47            Kian Mandiri Farm            Zainal Arifin   
70  2023   19  87.35                 Siluk Coffee                   Marjan   
71  2023   20  87.35  Suket Mountain/Gunung Suket               Supriyanto   
72  2023   21  87.32           Rukun Saudara Farm       Abdul Aziz Thohari   
73  2023   22  87.24              Empus Reje Guru              Ferri Padli   

        Region          Process     Variety_1        Variety_2 Variety_3  \
0         Aceh            Honey         Ateng           Gayo 1       P88   
1   Jawa Barat          Natural  Sigararutang              NaN       NaN   
2         Aceh           Washed         Ateng           Gayo 1       P88   
3   Jawa Timur          Natural          USDA  Colombia Brazil       NaN   
4   Jawa Barat          Natural  Sigararutang          Kartika     S-795   
..         ...              ...           ...              ...       ...   
69        Aceh   Mossto Natural         Ateng             P-88    Typica   
70        Aceh       Wet Hulled         Ateng              NaN       NaN   
71  Jawa Timur          Natural       Bourbon           Typica       NaN   
72       Jambi  Honey Caramella  Sigararutang       Andungsari       NaN   
73        Aceh          Natural   Red Bourbon              NaN       NaN   

   Variety_4  Weight (Lbs)  High Bid (US$)  Total Value (US$)  \
0        NaN        462.97           80.00           37037.60   
1        NaN        410.06           65.20           26735.91   
2        NaN        577.61           69.10           39912.85   
3        NaN        791.46           35.20           27859.39   
4        NaN        454.15           25.10           11399.17   
..       ...           ...             ...                ...   
69       NaN        621.44           10.20            6338.69   
70       NaN        898.03           13.60           12213.21   
71       NaN        908.30            6.25            5676.88   
72       NaN        588.37           16.40            9649.27   
73       NaN        577.39            8.45            4878.95   

                                                Buyer  
0                          Wataru for YAMATOYA COFFEE  
1                          Proud Mary Coffee Roasters  
2                           Terarosa (Haksan Co. Ltd)  
3                                      SUPREMO COFFEE  
4   Latorre&Dutch (China) for Cut Hand Group（剁手咖啡群...  
..                                                ...  
69                                      Campos Coffee  
70                              Coffee Beyond Borders  
71                                 Wataru & Co., Ltd.  
72                                   Solberg & Hansen  
73  Cuppers of Korea, Woojongho Coffee, Imbarista,...  

[74 rows x 15 columns]

## Save the Final Dataset to .csv Format

In [97]:
PATH = '/home/er_bim/Indonesia_COE/data/processed/cleaned/'

coe_indonesia.to_csv(PATH + 'coe_indonesia' + '.csv')